File: lib/legacy_rails/app/models/media_piece.rb
class MediaPiece < ApplicationRecord
  before_save :default_values

  belongs_to :ad_category
  belongs_to :media_piece_type
  belongs_to :marketer
  has_many :offers
  has_many :media_runs

  validates :banner,  presence: true
  validates :title,  presence: true
  validates :body_copy,  presence: true
  validates :display_url,  presence: true
  validates :jump_url,  presence: true
  validates :ad_category_id,  presence: true

  has_one_attached :banner

  
  # has_attached_file :resource
  # has_attached_file :adget_banner
  # validates_attachment_content_type :resource, content_type: /\Aimage\/.*\Z/

  def default_values
    self.active = true if self.active.nil?
    self.media_piece_type_id ||= 1
    self.marketer_id ||= 1
  end
end

File: lib/qlarius/legacy/media_piece.ex
defmodule Qlarius.Legacy.MediaPiece do
  use Ecto.Schema
  import Ecto.Changeset
  use Waffle.Ecto.Schema

  alias Qlarius.Legacy.{MediaPieceType, AdCategory}

  @primary_key {:id, :id, autogenerate: true}
  @timestamps_opts [type: :naive_datetime, inserted_at: :created_at, updated_at: :updated_at]

  schema "media_pieces" do
    field :title, :string
    field :body_copy, :string
    field :display_url, :string
    field :jump_url, :string
    field :active, :boolean
    field :marketer_id, :integer
    field :duration, :integer
    field :banner_image, QlariusWeb.ThreeTapBanner.Type

    belongs_to :media_piece_type, MediaPieceType
    belongs_to :ad_category, AdCategory

    has_many :media_runs, Qlarius.Legacy.MediaRun
    has_many :offers, through: [:media_runs, :offers]

    timestamps()
  end

  def changeset(media_piece, attrs) do
    media_piece
    |> cast(attrs, [
      :title,
      :body_copy,
      :display_url,
      :jump_url,
      :active,
      :marketer_id,
      :media_piece_type_id,
      :ad_category_id,
      :duration,
      :banner_image
    ])
    |> validate_required([
      :title,
      :display_url,
      :jump_url,
      :media_piece_type_id,
      :ad_category_id,
      :marketer_id,
      :active
    ])
    |> foreign_key_constraint(:media_piece_type_id)
    |> foreign_key_constraint(:ad_category_id)
    |> foreign_key_constraint(:marketer_id)
    |> cast_attachments(attrs, [:banner_image])
  end
end
